src/CompoundV3YieldOracle.sol

```diff
@@ -0,0 +1,237 @@
+
+        /// go from writeIdx to end of initialized array
+        for (uint256 i = writeIdx; i < snapshots.length; i++) {
+            SupplyRateSnapshot memory snapshot = snapshots[i];
+            /// break from loop if snapshot is not initialized
+            if (snapshot.timestamp == 0) {
+                break;
+            }
+            _updateYieldInfo(yieldInfo, snapshot);
+        }
+
+        /// go from 0 to writeIdx - 1
+        for (uint256 i = 0; i < writeIdx; i++) {
+            SupplyRateSnapshot memory snapshot = snapshots[i];
+            _updateYieldInfo(yieldInfo, snapshot);
+        }
+
+        return yieldInfo.weightedRateAcc / yieldInfo.totalDelta;
+    }
+
+    /// @inheritdoc IYieldOracle
+    function scale() public pure returns (uint8) {
+        return 18;
+    }
+
+    /**
+     * //////////// Keep3r ////////////
+     */
+
+    function work() external validateAndPayKeeper(msg.sender) {
+        _latchYieldForRefreshTokens();
+    }
+
+    /**
+     * //////////// ICompoundV3YieldOracle //////////////
+     */
+
+    /// @inheritdoc ICompoundV3YieldOracle
+    function setCometAddress(address baseAssetErc20, address comet)
+        public
+        requiresAdmin(msg.sender)
+        returns (address, address)
+    {
+        if (baseAssetErc20 == address(0)) {
+            revert InvalidTokenAddress();
+        }
+        if (comet == address(0)) {
+            revert InvalidCometAddress();
+        }
+
+        _updateTokenRefreshList(baseAssetErc20);
+
+        tokenAddressToComet[IERC20(baseAssetErc20)] = IComet(comet);
+        _setCometSnapShotBufferSize(baseAssetErc20, DEFAULT_SNAPSHOT_ARRAY_SIZE);
+
+        emit CometSet(baseAssetErc20, comet);
+        return (baseAssetErc20, comet);
+    }
+
+    /// @inheritdoc ICompoundV3YieldOracle
+    function latchCometRate(address token) external requiresAdmin(msg.sender) returns (uint256) {
+        return _latchSupplyRate(token);
+    }
+
+    /// @inheritdoc ICompoundV3YieldOracle
+    function latchRatesForRegisteredTokens() external requiresAdmin(msg.sender) {
+        _latchYieldForRefreshTokens();
+    }
+
+    /// @inheritdoc ICompoundV3YieldOracle
+    function getCometSnapshots(address token) public view returns (uint16 idx, SupplyRateSnapshot[] memory snapshots) {
+        IERC20 _token = IERC20(token);
+        idx = tokenToSnapshotWriteIndex[_token];
+        snapshots = tokenToSnapshotArray[_token];
+    }
+
+    /// @inheritdoc ICompoundV3YieldOracle
+    function setCometSnapshotBufferSize(address token, uint16 newSize)
+        external
+        requiresAdmin(msg.sender)
+        returns (uint16)
+    {
+        return _setCometSnapShotBufferSize(token, newSize);
+    }
+
+    /**
+     * /////////// Internal ///////////
+     */
+
+    function _setCometSnapShotBufferSize(address token, uint16 newSize) internal returns (uint16) {
+        SupplyRateSnapshot[] storage snapshots = tokenToSnapshotArray[IERC20(token)];
```

Comment:
SetCometSnapshotBufferSize is only accessible to the admin. Idea was to leave the size as a runtime parameter. 

we could (cheaply) enforce a max size on the snapshot buffer at a fixed constant (10? 20?) 

or allow for reduction in the size of the array, but this would be a bit more complex.

Reply:
Cheap gas wise is good, how big does the buffer array need to be? I think a deque with pop left is also a way to do this here.

Reply:
> how big does the buffer array need to be?

it really depends on what level of resolution we'd like for the yield. looking at black scholes, it seems that the risk free rate has a lower impact on the overall pricing of the option relative to other parameters, so maybe we don't need a very high resolution. maybe a 5d lookback with a 1d sampling period?

Reply:
I think that'd be fine.